File: rest-api/src/main/java/space/shefer/receipt/rest/service/UserProfileService.java
package space.shefer.receipt.rest.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import space.shefer.receipt.rest.entity.UserProfile;
import space.shefer.receipt.rest.repository.UserProfileRepository;

import java.util.UUID;

@Service
public class UserProfileService {

  private final UserProfileRepository userProfileRepository;

  @Autowired
  public UserProfileService(UserProfileRepository userProfileRepository) {
    this.userProfileRepository = userProfileRepository;
  }

  public UserProfile createOrUpdate(String phone, String password) {
    UserProfile userProfile = userProfileRepository.getByPhone(phone);
    if (userProfile == null) {
      UserProfile resultUser = new UserProfile();
      resultUser.setPassword(password);
      resultUser.setPhone(phone);
      resultUser.setAccessToken(UUID.randomUUID().toString());
      return userProfileRepository.save(resultUser);
    }
    if (!userProfile.getPassword().equals(password)) {
      userProfile.setPassword(password);
      userProfile.setPhone(phone);
      return userProfileRepository.save(userProfile);
    }
    return userProfile;
  }

  public String getAccessToken(String phone) {
    UserProfile userProfile = userProfileRepository.getByPhone(phone);
    if (userProfile != null) {
      return userProfile.getAccessToken();
    }
    return null;
  }

}

File: telegram-bot/src/main/kotlin/space/shefer/receipt/bot/service/UserProfileService.kt
package space.shefer.receipt.bot.service

import org.springframework.stereotype.Service
import org.telegram.telegrambots.meta.api.objects.Contact
import space.shefer.receipt.bot.model.UserProfile
import space.shefer.receipt.bot.repository.UserProfileRepository

@Service
class UserProfileService(
        private val userProfileRepository: UserProfileRepository
) {

    fun getOrCreate(contact: Contact): UserProfile {
        return userProfileRepository.findByPhoneNumber(contact.phoneNumber)
                ?: userProfileRepository.save(UserProfile().also {
                    it.firstName = contact.firstName
                    it.lastName = contact.lastName
                    it.phoneNumber = contact.phoneNumber
                })
    }

}
